### PR TITLE
Bigweek

### PR DESCRIPTION
**Pull request by ConorMcPersonal**
Some sorting out of the object model for Slot/SlotMonitor - in particular the slot doesn't care what temperature it used to be, only the monitor cares about that (same for the other old_ values).  Other simplifications so we don't need to pass increment around etc.  Plus a constructor-style function for a SlotState object.

This all enables multiple slots on the screen (4, as per the wiki) without too much extra code in total.

Please review only for now, will be working on this through the week.

### CODE REVIEW

**Comment by ConorMcPersonal** [comment]
Wondering about dropping the reverse polish notation (scandal I know)

Could just move to immediate single-key commands e.g.
"W" puts white bread in first available slot
"2" pops slot 2

Would make this less crap but more playable.  Thoughts?

File: SplitFiles/slot_monitor.c [hunk 1]
```diff
@@ -19,54 +19,57 @@
 #include "util.h"
 
 
-int plot_value(const unsigned int slot, const int value, const unsigned int task,
-    const int range, const int max, bool increment, const int old_value)
+int plot_value(SlotMonitor* slotMon, const int value, const unsigned int task,
```
Comment (ConorMcPersonal) [inline SplitFiles/slot_monitor.c:36]: Wondering if there's some realtime programmer thing I'm not getting here - do you prefer passing by value for performance reasons?  I think passing compound objects means shorter argument lists and more readable code but wonder if there's a performance hit you're aware of/worrying about that I am not?

Comment (mcmenaminadrian) [reply SplitFiles/slot_monitor.c:36]: No it's more those objects are yours and I'm not fully across them, I doubt there is much of a performance difference in statically compiled C code either way, so cleaner code is what should win. Just tell me how it should be cleaned up!

Comment (ConorMcPersonal) [reply SplitFiles/slot_monitor.c:36]: Think I have made it "cleaner" here but I was worrying I'm doing something really annoying!

File: SplitFiles/bread.c [hunk 2]
```diff
@@ -0,0 +1,42 @@
+//////////////////////////////////////////////////////////
```
Comment (ConorMcPersonal) [inline SplitFiles/bread.c:1]: Keeping this in here - think it might be useful

File: SplitFiles/bread.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// 1.) Forward declaration: Here is the name of the structure
+// but member-details are omitted.
+struct BreadBinStruct;
+
+// 2.) typedef of the structure
+// Still no details on the members.
+typedef struct BreadBinStruct  BreadBin;
+
+struct BreadStateStruct {
+  int               temperature;
```
Comment (ConorMcPersonal) [inline SplitFiles/bread.h:15]: Note this is not yet integrated

File: SplitFiles/control.c [hunk 4]
```diff
@@ -165,7 +162,7 @@ void execute_command(ControlBuffer *ctrlBuff, GameParameters* params) {
         //Push some bread to a slot
         d = buffer_pop(ctrlBuff);
         e = buffer_pop(ctrlBuff);
-        if (d >= '0' && d <= '9' && (e == 'W' || e == 'B' || e == 'G'))  {
+        if (d >= '0' && d <= '9' && (e == 'W' || e == 'B' || e == 'G' || e == 'C'))  {
```
Comment (ConorMcPersonal) [inline SplitFiles/control.c:165]: Getting Ciabatta to toast needed code changing in at least three places - in a data-driven approach we want to define it once and have it all just work

File: SplitFiles/control.c [hunk 5]
```diff
@@ -216,11 +213,22 @@ void execute_command(ControlBuffer *ctrlBuff, GameParameters* params) {
...
+                    buffer_push(d, ctrlBuff);
+                    buffer_push('x', ctrlBuff);
+                    return;
+                }
+                break;
             }
             new_slice->toastedness = 0;
-            new_slice->moisture = new_slice->old_moisture;
-            //draw the initial moisture setting
-            draw_moisture(d, new_slice->old_moisture, MAX_RANGE);
```
Comment (ConorMcPersonal) [inline SplitFiles/control.c:223]: No longer needed

**Comment by ConorMcPersonal** [comment]
I'd like to get this merged if possible - if you get a chance

File: SplitFiles/control.c [hunk 6]
```diff
@@ -175,7 +172,7 @@ void execute_command(ControlBuffer *ctrlBuff, GameParameters* params) {
                 case 'W':
                 if (fetch_bread(e)) {
                     // white bread is driest and quickest to toast
-                    new_slice->old_moisture = 64 + rand()%48;
+                    new_slice->moisture = 64 + rand()%48;
```
Comment (mcmenaminadrian) [inline SplitFiles/control.c:175]: I think this should be old_moisture for graphing/charting purposes - moisture is then copied from it. Can be the other way around of course

**Comment by mcmenaminadrian** [comment]
Comment in the code says you want it merged ASAP: which came first the
comment there or the 'let me get on with it' here?

**Comment by mcmenaminadrian** [comment]
ah right, I see I was reading these things back-to-front and the please
review to merge thing comes later. Ok will try to get this done today.

File: SplitFiles/control.c (hunk 6 above)
Comment (mcmenaminadrian) [reply SplitFiles/control.c:175]: ah, you have done it the other way round I see. No problem then

File: SplitFiles/control.c [hunk 7]
```diff
@@ -191,7 +188,7 @@ void execute_command(ControlBuffer *ctrlBuff, GameParameters* params) {
                 case 'B':
                 if (fetch_bread(e)) {
                     // brown takes longer
-                    new_slice->old_moisture = 96 + rand()%64;
+                    new_slice->moisture = 96 + rand()%64;
                     new_slice->thermalMass = 82;
```
Comment (mcmenaminadrian) [inline SplitFiles/control.c:192]: For smoother plotting these thermal mass numbers should probably be multiples of 16. Doesn't really matter much though.

File: SplitFiles/slot_monitor.c [hunk 8]
```diff
@@ -33,55 +33,58 @@ void draw_moisture(const int slot, const int moisture, const int max)
...
-        x = 16 + old_value / step_value;
-        if (increment) {
-            for (i = old_value; i < value; i+= step_value) {
-                for (j = 0; j < 3; j++) {
-                    if (x > 16) { // avoid negative charting
-                        *zx_pxy2saddr(x, offset + j) |= zx_px2bitmask(x);
+    if (value != old_value) {
+        bool increment = (value > old_value);
+        if (value <= max && old_value >= 0) {
+            const unsigned int offset = (slotMon->yBase - 1) * 8 + task * 4;
```
Comment (mcmenaminadrian) [inline SplitFiles/slot_monitor.c:42]: Lots of magic numbers. I know this was me!

File: SplitFiles/slot_monitor.c [hunk 9]
```diff
@@ -33,55 +33,58 @@ void draw_moisture(const int slot, const int moisture, const int max)
...
-        slot->bread->old_moisture = plot_value(slot->slotNumber, slot->bread->moisture, 2, 64,
-            MAX_RANGE, false, slot->bread->old_moisture);
-        slot->bread->old_toastedness = plot_value(slot->slotNumber, slot->bread->toastedness, 3, 64,
-            MAX_RANGE, increment, slot->bread->old_toastedness);
-    } else if (wipe_out == false){
+        slotMon->bread_old_moisture = plot_value(slotMon, slot->bread->moisture, 2, 64,
+            256, slotMon->bread_old_moisture);
+        slotMon->bread_old_toastedness = plot_value(slotMon, slot->bread->toastedness, 3, 64,
+            256, slotMon->bread_old_toastedness);
+    } else if (slotMon->bread_old_moisture != 0 || slotMon->bread_old_toastedness != 0) {
```
Comment (mcmenaminadrian) [inline SplitFiles/slot_monitor.c:82]: Again many more magic numbers

**Review by mcmenaminadrian — APPROVED** [review]
No reason not to merge though I have identified a few places where we could make improvements, but they could be the next task(s). To be honest it's starting to feel like a 'crap' game now, so I think we are making real progress.

File: SplitFiles/control.c (hunk 6 above)
Comment (ConorMcPersonal) [reply SplitFiles/control.c:175]: Yeah that was the object model change - old_moisture is something the slot monitor needs to know about, but the bread shouldn't care - so I moved it over.  Might be me being an OO wanker, but this was grating on my conceptual model!!

File: SplitFiles/slot_monitor.c (hunk 9 above)
Comment (ConorMcPersonal) [reply SplitFiles/slot_monitor.c:82]: Yeah we're passing 64 and 256 on every call to plot_value; we should replace them with a #define and simplify the argument list

**Pull request merged**